// @tetherto/embed-llamacpp/addon.js
const binding = require('./binding')

// TODO(jesusmb1995): Bind to summarized word embeddings and utilities.
/// An interface between Bare addon in C++ and JS runtime.
class BertInterface {
  /**
   *
   * @param {Object} configurationParams - all the required configuration for inference setup
   * @param {Function} outputCb - to be called on any inference event ( started, new output, error, etc )
   * @param {Function} transitionCb - to be called on addon state changes (LISTENING, IDLE, STOPPED, etc )
   */
  constructor (configurationParams, outputCb, transitionCb = null) {
    this._handle = binding.createInstance(this, configurationParams, outputCb, transitionCb)
  } ///

  /**
   * Cancel a inference process by jobId, if no jobId is provided it cancel the whole queue
   */
  async cancel (jobId) {
    binding.cancel(this._handle, jobId)
  }

  /**
   * Adds new input to the processing queue
   * @param {Object} data
   * @param {String} data.type
   * @param {String} data.input
   * @returns {Promise<Number>} - job ID
   */
  async append (data) {
    return binding.append(this._handle, data)
  }

  /**
   * Addon process status
   * @returns {Promise<String>}
   */
  async status () {
    return binding.status(this._handle)
  }

  /**
   * Stops addon process and clears resources (including memory).
   */
  async destroyInstance () {
    if (!this._handle) return
    binding.destroyInstance(this._handle)
    this._handle = null
  }

  /**
   * Loads model weights
   * @param {Object} data
   * @param {String} data.filename
   * @param {Buffer} data.contents
   * @param {Promise<Boolean>} data.completed
   */
  async loadWeights (data) {
    return binding.loadWeights(this._handle, data)
  }

  /**
   * Activates the model to start processing the queue
   */
  async activate () {
    return binding.activate(this._handle)
  }

  /**
   * Resets the model state
   */
  async reset () {
    return binding.reset(this._handle)
  }

  async unload () {
    if (!this._handle) return
    binding.destroyInstance(this._handle)
    this._handle = null
  }
}

module.exports = {
  BertInterface
}
