// @tetherto/embed-llamacpp/index.js
'use strict'

const path = require('bare-path')
const BaseInference = require('@tetherto/infer-base/WeightsProvider/BaseInference')
const WeightsProvider = require('@tetherto/infer-base/WeightsProvider/WeightsProvider')
const { BertInterface } = require('./addon')

const END_OF_INPUT = 'end of job'

/**
 * GGML client implementation for BERT GTE model
 */
class GGMLBert extends BaseInference {
  /**
   * Creates an instance of GGMLBert.
   * @constructor
   * @param {Object} params - arguments for model setup
   * @param {Object} args arguments for inference setup
   * @param {Object} config - environment specific inference setup configuration
   */
  constructor (
    { opts = {}, loader, logger = null, diskPath, modelName },
    config
  ) {
    super({ logger, opts })
    this._config = config
    this._diskPath = diskPath
    this._modelName = modelName
    this.weightsProvider = new WeightsProvider(loader, this.logger)
  }

  async _load (closeLoader = false, reportProgressCallback) {
    this.logger.info('Starting model load')
    await this.downloadWeights(reportProgressCallback, { closeLoader })

    const configurationParams = {
      path: path.join(this._diskPath, this._modelName),
      config: this._config
    }

    this.addon = this._createAddon(configurationParams)

    this.logger.info('Activating addon')
    await this.addon.activate()

    this.logger.info('Model load completed successfully')
  }

  /**
   * Download the model weight files and return the local path to the primary file.
   * @param {ProgressReportCallback} [onDownloadProgress] - Callback invoked with bytes downloaded
   * @param {Object} opts - Options for the download
   * @param {boolean} opts.closeLoader - Whether to close the loader when done
   * @returns {Promise<{filePath: string, completed: boolean, error: boolean}[]>} Local file path for the model weights
   */
  async _downloadWeights (onDownloadProgress, opts) {
    return await this.weightsProvider.downloadFiles(
      [this._modelName],
      this._diskPath,
      {
        closeLoader: opts.closeLoader,
        onDownloadProgress
      }
    )
  }

  async _runInternal (text) {
    this.logger.info('Starting inference embeddings for text:', text)

    const jobId = await this.addon.append({ type: 'text', input: text })

    const response = this._createResponse(jobId)

    await this.addon.append({ type: END_OF_INPUT })

    return response
  }

  /**
   * Instantiate the native addon with the given parameters.
   * @param {Object} configurationParams - Configuration parameters for the addon
   * @param {string} configurationParams.path - Local file or directory path
   * @param {Object} configurationParams.settings - Bert-specific settings
   * @returns {Addon} The instantiated addon interface
   */
  _createAddon (configurationParams) {
    this.logger.info(
      'Creating Bert interface with configuration:',
      configurationParams
    )
    return new BertInterface(
      configurationParams,
      this._outputCallback.bind(this),
      console.log
    )
  }
}

module.exports = GGMLBert
