// @tetherto/qvac-lib-inference-addon-mlc-base/lib/error.js
'use strict'

const { QvacErrorBase, addCodes } = require('@tetherto/qvac-lib-error-base')
const { name, version } = require('../package.json')

// Define error codes specific to this library - range 5001-6000
const ERR_CODES = Object.freeze({
  LOADING_WEIGHTS_FAILED: 5001,
  DOWNLOADING_WEIGHTS_FAILED: 5002,
  UNLOADING_WEIGHTS_FAILED: 5003
})

// Register error definitions
addCodes(
  {
    [ERR_CODES.LOADING_WEIGHTS_FAILED]: {
      name: 'LOADING_WEIGHTS_FAILED',
      message: msg => `Loading weights failed: ${msg}`
    },
    [ERR_CODES.DOWNLOADING_WEIGHTS_FAILED]: {
      name: 'DOWNLOADING_WEIGHTS_FAILED',
      message: msg => `Downloading weights failed: ${msg}`
    },
    [ERR_CODES.UNLOADING_WEIGHTS_FAILED]: {
      name: 'UNLOADING_WEIGHTS_FAILED',
      message: msg => `Unloading weights failed: ${msg}`
    }
  },
  { name, version }
)

// Create a specialized error class for the inference addon mlc base library
class QvacInferenceAddonMlcBaseError extends QvacErrorBase {}

module.exports = {
  QvacInferenceAddonMlcBaseError,
  ERR_CODES
}
